src/callbacks_step/save_solution_dg.jl

```diff
@@ -103,6 +103,65 @@ function save_solution_file(u, time, dt, timestep,
     n_vars = size(data, 1)
   end
 
+  if HDF5.has_parallel()
+    save_solution_file_parallel(data, time, dt, timestep, n_vars, mesh, equations, dg, cache, solution_variables, filename, element_variables)
+  else
+    save_solution_file_on_root(data, time, dt, timestep, n_vars, mesh, equations, dg, cache, solution_variables, filename, element_variables)
+  end
+end
+
+
+function save_solution_file_parallel(data, time, dt, timestep, n_vars,
+                                     mesh::Union{ParallelTreeMesh, ParallelP4estMesh}, equations, dg::DG, cache,
+                                     solution_variables, filename, element_variables=Dict{Symbol,Any}())
+  # Open file using parallel HDF5 (clobber existing content)
+  h5open(filename, "w", mpi_comm()) do file
+    # Add context information as attributes
+    attributes(file)["ndims"] = ndims(mesh)
+    attributes(file)["equations"] = get_name(equations)
+    attributes(file)["polydeg"] = polydeg(dg)
+    attributes(file)["n_vars"] = n_vars
+    attributes(file)["n_elements"] = nelementsglobal(dg, cache)
+    attributes(file)["mesh_type"] = get_name(mesh)
+    attributes(file)["mesh_file"] = splitdir(mesh.current_filename)[2]
+    attributes(file)["time"] = convert(Float64, time) # Ensure that `time` is written as a double precision scalar
+    attributes(file)["dt"] = convert(Float64, dt) # Ensure that `dt` is written as a double precision scalar
+    attributes(file)["timestep"] = timestep
+
+    # Store each variable of the solution data
+    for v in 1:n_vars
+      # Need to create dataset explicitly in parallel case
+      var = create_dataset(file, "/variables_$v", datatype(eltype(data)), dataspace((ndofsglobal(mesh, dg, cache),)))
+
+      # Write data of each process in slices
+      slice = mpi_rank()*ndofs(mesh, dg, cache) + 1:(mpi_rank() + 1)*ndofs(mesh, dg, cache)
+      # Convert to 1D array
+      var[slice] = vec(data[v, .., :])
+      # Add variable name as attribute
+      attributes(var)["name"] = varnames(solution_variables, equations)[v]
+    end
+
+    # Store element variables
+    for (v, (key, element_variable)) in enumerate(element_variables)
+      # Need to create dataset explicitly in parallel case
+      var = create_dataset(file, "/element_variables_$v", datatype(eltype(element_variable)), dataspace((nelementsglobal(dg, cache),)))
+
+      # Write data of each process in slices
+      slice = mpi_rank()*nelements(dg, cache) + 1:(mpi_rank() + 1)*nelements(dg, cache)
```

Comment:
Maybe I do not fully understand this part, but wouldn't this only work if `nelements(dg, cache)` is the same on each rank?

Reply:
Yes, you are absolutely right. Thanks, I missed that. We need to compute the slice that the current rank writes to from `cache.mpi_cache.n_elements_by_rank`. I will fix that.
Do we need the same for `ndofs` (but there is no `n_dofs_by_rank` in the `mpi_cache` as far as I can see)?

Reply:
Yes, I believe so.

Reply:
I committed a solution. Please have a look if this looks fine to you.

Reply:
LGTM, thanks!